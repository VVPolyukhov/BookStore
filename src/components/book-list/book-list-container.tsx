import React, { Component } from 'react'
import { bindActionCreators, Dispatch } from 'redux'
import { connect } from 'react-redux'
import orderBy from 'lodash/orderBy'

import BookList from './book-list'
import Spinner from '../spinner'
import ErrorIndicator from '../error-indicator'
import { withBookstoreService } from '../hoc'

import { compose } from '../../utils'

import { fetchBooks } from '../../actions/book-list'
import { ReducerType, BookType } from '../../types'

type MapStateToPropsType = {
    books: Array<BookType>
    loading: boolean
    error: boolean | null
}

type MapDispatchToPropsType = {
    fetchBooks: () => void
}

type PropsType = MapStateToPropsType & MapDispatchToPropsType

class BookListContainer extends Component<PropsType> {

    componentDidMount() {
        this.props.fetchBooks()
    }

    render() {
        const { books, loading, error } = this.props

        if (loading)
            return <Spinner />

        if (error)
            return <ErrorIndicator />

        return <BookList books={books} />
    }
}

const sortBy = (books : Array<BookType>, filterBy : string) => {
    switch (filterBy) {
        case 'all':
            return books
        case 'popularity':
            return orderBy(books, 'rating', 'desc')
        case 'low-price':
            return orderBy(books, 'price', 'asc')
        case 'high-price':
            return orderBy(books, 'price', 'desc')
        case 'bookName':
            return orderBy(books, 'title', 'asc')
        case 'author':
            return orderBy(books, 'author', 'asc')
        default:
            return books
    }
}

const searchBooks = (books : Array<BookType>, term : string) => {
    if (term.length === 0) {
        return books;
    }
    return books.filter(book => {
        return (book.title.toLowerCase().indexOf(term) > -1 ||
            book.author.toLowerCase().indexOf(term) > -1);
    });
}

const mapStateToProps = ({ bookList: { books, loading, error }, 
                           filter:   { term, filterBy } } : ReducerType) 
                        : MapStateToPropsType => {
    return { 
        books: books &&
                sortBy(
                    searchBooks(books, term.toLowerCase()), 
                    filterBy),
        loading, 
        error
    }
}

const mapDispatchToProps = (dispatch: Dispatch, { bookstoreService } : any) 
                            : MapDispatchToPropsType => {
    return bindActionCreators({
        fetchBooks: fetchBooks(bookstoreService)
    }, dispatch);
};

export default compose(
    withBookstoreService(),
    connect<MapStateToPropsType, MapDispatchToPropsType, {}, ReducerType>
    (mapStateToProps, mapDispatchToProps)
)(BookListContainer)