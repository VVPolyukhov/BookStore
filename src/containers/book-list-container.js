import React, { Component } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import orderBy from 'lodash/orderBy'

import BookList from '../components/book-list'
import Spinner from '../components/spinner'
import ErrorIndicator from '../components/error-indicator'
import { withBookstoreService } from '../components/hoc'

import { compose } from '../utils'

import { fetchBooks } from '../actions/book-list'

class BookListContainer extends Component {

    componentDidMount() {
        this.props.fetchBooks()
    }

    render() {
        const { books, loading, error, onAddedToCart } = this.props

        if (loading)
            return <Spinner />

        if (error)
            return <ErrorIndicator />

        return <BookList books={books}
                         onAddedToCart={onAddedToCart} />
    }
}

const sortBy = (books, filterBy) => {
    switch (filterBy) {
        case 'all':
            return books
        case 'popularity':
            return orderBy(books, 'rating', 'desc')
        case 'low-price':
            return orderBy(books, 'price', 'asc')
        case 'high-price':
            return orderBy(books, 'price', 'desc')
        case 'bookName':
            return orderBy(books, 'title', 'asc')
        case 'author':
            return orderBy(books, 'author', 'asc')
        default:
            return books
    }
}

const searchBooks = (books, term) => {
    if (term.length === 0) {
        return books;
    }
    return books.filter(book => {
        return (book.title.toLowerCase().indexOf(term) > -1 ||
            book.author.toLowerCase().indexOf(term) > -1);
    });
}

const mapStateToProps = ({ bookList: { books, loading, error }, 
                           filter:   { term, filterBy } }) => {
    return { 
        books: books &&
                sortBy(
                    searchBooks(books, term.toLowerCase()), 
                    filterBy),
        loading, 
        error
    }
}

const mapDispatchToProps = (dispatch, { bookstoreService }) => {
    return bindActionCreators({
        fetchBooks: fetchBooks(bookstoreService)
    }, dispatch);
};

export default compose(
    withBookstoreService(),
    connect(mapStateToProps, mapDispatchToProps)
)(BookListContainer)