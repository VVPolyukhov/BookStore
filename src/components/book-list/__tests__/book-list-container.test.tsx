import React from 'react'
import { BookListContainer, PropsType } from '../book-list-container'
import { shallow } from 'enzyme'

describe('BookListContainer:', () => {
  const props: PropsType = { 
      books: [],
      loading: false,
      error: null,
      fetchBooks: () => {}
  }

  /*
  describe('BookListContainer - initial', () => { 
    const mockFetchBooks = jest.fn()
    const nextProps: PropsType = {
        ...props, 
        fetchBooks: mockFetchBooks
    }

    const Wrapper = shallow(<BookListContainer {...nextProps} />)
    it('renders properly', () => {
      expect(Wrapper).toMatchSnapshot()
    })
    it('dispatches the `fetchBooks()` method it receives from props', () => {
      expect(mockFetchBooks).toHaveBeenCalled()
    })
  })
  */
 
  describe('BookListContainer - loading', () => {
    const nextProps = {
      ...props,
      loading: true,
    }

    const Wrapper = shallow(<BookListContainer {...nextProps} />)
    it('renders properly', () => {
      expect(Wrapper).toMatchSnapshot()
    })
    it('render with loading', () => {
      expect(Wrapper.find('Spinner')).toHaveLength(1) 
      expect(Wrapper.find('ErrorIndicator')).toHaveLength(0)
      expect(Wrapper.find('BookList')).toHaveLength(0)
    })
  })

  describe('BookListContainer - error', () => {
    const nextProps = {
      ...props,
      error: true,
    }
    
    const Wrapper = shallow(<BookListContainer {...nextProps} />)
    it('renders properly', () => {
      expect(Wrapper).toMatchSnapshot()
    })
    it('render with error', () => {
      expect(Wrapper.find('Spinner')).toHaveLength(0) 
      expect(Wrapper.find('ErrorIndicator')).toHaveLength(1)
      expect(Wrapper.find('BookList')).toHaveLength(0)
    })
  })

  describe('BookListContainer renders BookList component', () => {
    const Wrapper = shallow(<BookListContainer {...props} />)
    it('renders properly', () => {
      expect(Wrapper).toMatchSnapshot()
    })
    it('render <BookList /> template', () => {
      expect(Wrapper.find('Spinner')).toHaveLength(0) 
      expect(Wrapper.find('ErrorIndicator')).toHaveLength(0)
      expect(Wrapper.find('BookList')).toHaveLength(1)
    })
  })
})