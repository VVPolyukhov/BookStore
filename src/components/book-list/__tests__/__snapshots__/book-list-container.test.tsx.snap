// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListContainer: BookListContainer - error renders properly 1`] = `<ErrorIndicator />`;

exports[`BookListContainer: BookListContainer - loading renders properly 1`] = `<Spinner />`;

exports[`BookListContainer: BookListContainer renders BookList component renders properly 1`] = `
<BookList
  books={Array []}
/>
`;
